File: backends/java/zappy-domain/src/test/java/com/zappymart/domain/catalogue/ProductSpecificationTest.java
package com.zappymart.domain.catalogue;

import org.junit.jupiter.api.Test;

import static com.zappymart.domain.builders.ProductBuilder.JEWELLERY;
import static com.zappymart.domain.builders.ProductBuilder.aProduct;
import static org.assertj.core.api.Assertions.assertThat;

class ProductSpecificationTest {

    private static final Product RING = aProduct().named("White Gold Plated Princess")
            .inCategory(JEWELLERY).withStock(0).build();

    private static final Product JACKET = aProduct().named("Mens Cotton Jacket").withStock(8).build();

    @Test
    void everyProductSatisfiesTheOpenSpecification() {
        assertThat(ProductSpecification.everyProduct().isSatisfiedBy(RING)).isTrue();
    }

    @Test
    void keepsOneCategory() {
        ProductSpecification jewellery = new ProductSpecification.InCategory("jewellery");

        assertThat(jewellery.isSatisfiedBy(RING)).isTrue();
        assertThat(jewellery.isSatisfiedBy(JACKET)).isFalse();
    }

    @Test
    void matchesANameWithoutRegardToCase() {
        ProductSpecification cotton = new ProductSpecification.NameContains("COTTON");

        assertThat(cotton.isSatisfiedBy(JACKET)).isTrue();
        assertThat(cotton.isSatisfiedBy(RING)).isFalse();
    }

    @Test
    void leavesOutWhatHasNoStock() {
        ProductSpecification inStock = new ProductSpecification.InStock();

        assertThat(inStock.isSatisfiedBy(RING)).isFalse();
        assertThat(inStock.isSatisfiedBy(JACKET)).isTrue();
    }

    @Test
    void composesPartsThatAllHaveToHold() {
        ProductSpecification both = ProductSpecification.everyProduct()
                .and(new ProductSpecification.InCategory("jewellery"))
                .and(new ProductSpecification.InStock());

        assertThat(both.isSatisfiedBy(RING)).isFalse();
        assertThat(both.and(ProductSpecification.everyProduct()).isSatisfiedBy(RING)).isFalse();
    }
}

File: backends/kotlin/zappy-domain/src/test/kotlin/nl/zappymart/domain/catalogue/ProductSpecificationTest.kt
package nl.zappymart.domain.catalogue

import nl.zappymart.domain.builders.aProduct
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class ProductSpecificationTest {

    private val ring = aProduct().named("Princess ring").inCategory("jewellery").withStock(0).build()

    private val jacket = aProduct().named("Cotton jacket").inCategory("mens-clothing").withStock(8).build()

    @Test
    fun `an empty specification keeps every product`() {
        assertThat(ProductSpecification.EVERYTHING.isSatisfiedBy(ring)).isTrue()
        assertThat(ProductSpecification.EVERYTHING.isSatisfiedBy(jacket)).isTrue()
    }

    @Test
    fun `a category filter keeps only that category`() {
        val specification = ProductSpecification.of("jewellery", null, null)
        assertThat(specification.isSatisfiedBy(ring)).isTrue()
        assertThat(specification.isSatisfiedBy(jacket)).isFalse()
    }

    @Test
    fun `a name filter ignores case`() {
        val specification = ProductSpecification.of(null, "COTTON", null)
        assertThat(specification.isSatisfiedBy(jacket)).isTrue()
        assertThat(specification.isSatisfiedBy(ring)).isFalse()
    }

    @Test
    fun `the in stock filter leaves out what has none`() {
        val specification = ProductSpecification.of(null, null, true)
        assertThat(specification.isSatisfiedBy(ring)).isFalse()
        assertThat(specification.isSatisfiedBy(jacket)).isTrue()
    }

    @Test
    fun `every part that is given has to hold`() {
        val specification = ProductSpecification.of("mens-clothing", "jacket", true)
        assertThat(specification.isSatisfiedBy(jacket)).isTrue()
        assertThat(ProductSpecification.of("jewellery", "jacket", true).isSatisfiedBy(jacket)).isFalse()
    }

    @Test
    fun `a blank name filter is left out`() {
        assertThat(ProductSpecification.of(null, "   ", null)).isEqualTo(ProductSpecification.EVERYTHING)
    }
}
